gui/src/deadliner.rs

```diff
@@ -64,7 +59,6 @@ pub struct Deadliner<'a> {
     textures: HashMap<&'a str, TextureHandle>,
 
     error_msg: String,
-    invalid_bg: bool,
```

Comment:
That's so much better architecture, related pieces of data are now tied together in one type.

But you've actually introduced a bug here. The App state needs to store the errors, cause remember Egui uses the `update` method to render the app and it runs it 60 times per second to achieve 60FPS smooth apps.

```diff
Background::FromDisk { location, .. } => {
            ui.horizontal(|ui| {
+                let mut is_valid = true;
                if ui.button("Open file…").clicked() {
                    if let Some(path) = rfd::FileDialog::new().pick_file() {
                        let new_location = path.display().to_string();

                        let file_name = get_file_name_from_path(&new_location);
                        let file_ext = file_name.split(".").collect::<Vec<&str>>().pop().unwrap();
                        let supported_file_ext = ["png", "gif", "jpg", "jpeg"];

                        if supported_file_ext.contains(&file_ext) {
                            *location = new_location;
                            is_valid = true;
                        } else {
                            is_valid = false;
                        }
                    }
                }

+                if !is_valid {
+                    ui.colored_label(Color32::from_rgb(255, 48, 48), "Not an Image");
+                } else if !location.is_empty() {
+                    ui.colored_label(
+                        Color32::from_rgba_unmultiplied(254, 216, 67, 200),
+                        get_file_name_from_path(location),
+                    );
+                }
+            });
+        }
```

So checking for a background if it was invalid or not in the update function would mean that if you've chosen an invalid file(not an image format) the error would show only for 1 frame in 1/60 second then It forgets about it after the first render.

The only place where we can persist a value across frames rendering is at the app state level.

Reply:
Here's a screen recording for illustrating what I mean

https://user-images.githubusercontent.com/72823042/172052572-e55d7ae9-9a99-43d2-92dd-5fc9f8f556e6.mp4

Reply:
Ah, good catch. I just pushed another commit that uses Egui's temporary widget state to persist the validity without polluting the application state. It's a little messier in the widget GUI code, but I think I like it better than putting the extra state in the app.

Still, I could easily go either way with this one.